### PR TITLE
Feature/sof 4311 - download_from_urls IO unit

### PR DESCRIPTION
**Pull request by AcylSilane**

### CODE REVIEW

File: schema/workflow/unit/io/download_from_urls.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "download-from-urls io schema",
+    "type": "object",
+    "allof": [
+        {
+            "$ref": "../../unit.json"
+        }
+    ],
+    "properties": {
+        "url": {
```
Comment (AcylSilane) [inline schema/workflow/unit/io/download_from_urls.json:12]: Replace with container information similar to 4309's PR (can probably be imported as a reference)

Comment (AcylSilane) [reply schema/workflow/unit/io/download_from_urls.json]: The container has been replaced with an "ObjectProperties" object

File: schema/workflow/unit/io/download_from_object_storage.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        "filename": {
+            "description": "name given to the file when it is downloaded",
+            "type": "string"
+        },
+        "objectData": {
+            "description": "information describing the file and where it was run",
+            "type": "object",
+            "properties": {
+                "CONTAINER": {
+                    "description": "Container that ran the job resulting in the file",
```
Comment (triplepoint) [inline schema/workflow/unit/io/download_from_object_storage.json:21]: I think this is misunderstanding what a container is.  This is container in the sense of "object store", like an aws s3 bucket: https://aws.amazon.com/s3/.  Not a container in the sense of a containerized process, as in docker.

File: schema/workflow/unit/io/download_from_object_storage.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                "NAME": {
+                    "description": "Path to where the job put the file",
+                    "type": "string"
+                },
+                "PROVIDER": {
+                    "description": "Provider of the computational resources",
+                    "type": "string"
+                },
+                "REGION": {
+                    "description": "Region of the cloud provider's server",
```
Comment (triplepoint) [inline schema/workflow/unit/io/download_from_object_storage.json:33]: This is specifically the region for the object container specified in CONTAINER.

**Review by triplepoint — CHANGES_REQUESTED** [review]

File: example/workflow/unit/io/download_from_object_storage.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    "objectData": {
+        "CONTAINER": "example-cluster-001",
+        "NAME": "/path/to/the/index.html",
+        "PROVIDER": "some-cloud-compute-provider",
+        "REGION": "us-west-1"
+    },
+    "flowchartId": "download-files",
+    "head": true,
+    "name": "download_files",
+    "type": "download_from_urls"
```
Comment (triplepoint) [inline example/workflow/unit/io/download_from_object_storage.json:13]: is this still accurate?  Or is it something like `download_from_object_storage`?

Comment (timurbazhirov) [reply example/workflow/unit/io/download_from_object_storage.json:13]: Type is io

File: schema/workflow/unit/io/download_from_object_storage.json (hunk 2 above)
Comment (AcylSilane) [reply schema/workflow/unit/io/download_from_object_storage.json:21]: Thanks for pointing that out. Fixed the description

File: schema/workflow/unit/io/download_from_object_storage.json (hunk 3 above)
Comment (AcylSilane) [reply schema/workflow/unit/io/download_from_object_storage.json:33]: Description here has been fixed

File: example/workflow/unit/io/download_from_object_storage.json (hunk 4 above)
Comment (AcylSilane) [reply example/workflow/unit/io/download_from_object_storage.json]: Fixed type to now be io

File: schema/workflow/unit/io.json [hunk 5]
```diff
@@ -24,7 +24,8 @@
         "source": {
             "enum": [
                 "api",
-                "db"
+                "db",
+                "object-storage"
```
Comment (timurbazhirov) [inline schema/workflow/unit/io.json:28]: object_storage

File: schema/workflow/unit/io/object_storage.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "object-storage io schema",
+    "type": "object",
+    "allof": [
+        {
+            "$ref": "../../unit.json"
+        }
+    ],
+    "properties": {
+        "filename": {
```
Comment (timurbazhirov) [inline schema/workflow/unit/io/object_storage.json:12]: Let's use the pathname and basename to be clear? If pathname is empty, we assume current directory

File: schema/workflow/unit/io.json (hunk 5 above)
Comment (AcylSilane) [reply schema/workflow/unit/io.json:28]: Updated to use snake_case here

File: schema/workflow/unit/io/object_storage.json (hunk 6 above)
Comment (AcylSilane) [reply schema/workflow/unit/io/object_storage.json]: Filename has been split into pathname and basename

File: schema/workflow/unit/io/object_storage.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    ],
+    "properties": {
+        "pathname": {
+            "description": "Name of the directory to contain the file when it is downloaded. If empty, the current directory is used",
+            "type": "string"
+        },
+        "basename": {
+            "description": "Basename the file should have when it is downloaded",
+            "type": "string"
+        },
```
Comment (timurbazhirov) [inline schema/workflow/unit/io/object_storage.json:19]: Let's isolate pathname/basename to its own shema for reusability

Comment (AcylSilane) [reply schema/workflow/unit/io/object_storage.json]: Isolated to its own schema

File: schema/workflow/unit/io/object_storage.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
+{
+    "schemaId": "workflow-unit-io-object-storage",
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "object_storage io schema",
+    "type": "object",
+    "allof": [
+        {
+            "$ref": "../../unit.json"
+        },
```
Comment (timurbazhirov) [inline schema/workflow/unit/io/object_storage.json:9]: This should not be included

File: schema/core/reusable/file_metadata.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        "basename": {
+            "description": "Basename of the file",
+            "type": "string"
+        },
+        "filetype": {
+            "description": "What kind of file this is, e.g. image / text",
+            "type": "string"
+        }
+    }
+}
```
Comment (timurbazhirov) [inline schema/core/reusable/file_metadata.json:19]: We should include this file in SOF-4309

Comment (AcylSilane) [reply schema/core/reusable/file_metadata.json]: Commits reverted and moved to 4309

File: schema/workflow/unit/io/object_storage.json (hunk 8 above)
Comment (AcylSilane) [reply schema/workflow/unit/io/object_storage.json]: Reference has been removed

**Review by timurbazhirov — APPROVED** [review]

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**